lib/reline/line_editor.rb

```diff
@@ -643,41 +648,120 @@ def add_dialog_proc(name, p, context = nil)
     str + (' ' * padding_width)
   end
 
-  private def render_each_dialog(dialog, cursor_column)
+  private def range_subtract(base_ranges, subtract_ranges)
+    indices = base_ranges.flat_map(&:to_a).uniq.sort - subtract_ranges.flat_map(&:to_a)
+    chunks = indices.chunk_while { |a, b| a + 1 == b }
+    chunks.map { |a| a.first...a.last + 1 }
+  end
+
+  private def dialog_range(dialog, dialog_y)
+    x_range = dialog.column...dialog.column + dialog.width
+    y_range = dialog_y + dialog.vertical_offset...dialog_y + dialog.vertical_offset + dialog.contents.size
+    [x_range, y_range]
+  end
+
+  private def render_dialog_changes(changes, cursor_column)
+    old_dialog_ranges = {}
+    new_dialog_ranges = {}
+    new_dialog_contents = {}
+    changes.each do |old_dialog, new_dialog|
+      if old_dialog.contents
+        x_range, y_range = dialog_range(old_dialog, @previous_rendered_dialog_y)
+        y_range.each do |y|
+          (old_dialog_ranges[y] ||= []) << x_range
+        end
+      end
+      if new_dialog.contents
+        x_range, y_range = dialog_range(new_dialog, @first_line_started_from + @started_from)
+        y_range.each do |y|
+          (new_dialog_ranges[y] ||= []) << x_range
+          (new_dialog_contents[y] ||= []) << [x_range, new_dialog.contents[y - y_range.begin]]
+        end
+      end
+    end
+    return if old_dialog_ranges.empty? && new_dialog_ranges.empty?
+    ranges_to_restore = {}
+    subtract_cache = {}
+    old_dialog_ranges.each do |y, old_x_ranges|
+      new_x_ranges = new_dialog_ranges[y] || []
+      ranges = subtract_cache[[old_x_ranges, new_x_ranges]] ||= range_subtract(old_x_ranges, new_x_ranges)
+      ranges_to_restore[y] = ranges if ranges.any?
+    end
+
+    if ranges_to_restore.any?
+      lines = whole_lines
+      prompt, _prompt_width, prompt_list = check_multiline_prompt(lines, force_recalc: true)
+      modified_lines = modify_lines(lines, force_recalc: true)
+      visual_lines = []
+      modified_lines.each_with_index { |l, i|
+        pr = prompt_list ? prompt_list[i] : prompt
+        vl, = split_by_width(pr + l, @screen_size.last)
+        vl.compact!
+        visual_lines.concat(vl)
+      }
+    end
+
+    Reline::IOGate.hide_cursor
+    ymin, ymax = (ranges_to_restore.keys + new_dialog_ranges.keys).minmax
+    dialog_y = @first_line_started_from + @started_from
+    cursor_y = dialog_y
+    scroll_down(ymax - cursor_y)
+    move_cursor_up(ymax - cursor_y)
+    (ymin..ymax).each do |y|
+      move_cursor_down(y - cursor_y)
+      cursor_y = y
+      new_x_ranges = new_dialog_ranges[y]
+      restore_ranges = ranges_to_restore[y]
+      if restore_ranges
+        line = visual_lines[y] || ''
+        restore_ranges.each do |range|
+          col = range.begin
+          width = range.end - range.begin
+          s = padding_space_with_escape_sequences(Reline::Unicode.take_range(line, col, width), width)
+          Reline::IOGate.move_cursor_column(col)
```

Comment:
One thing that always puzzles me is all the cursor operations. Why can't we update the cursor column only once (e.g. outside this loop), but have to move it so many times 🤔

Reply:
In terminal programming, we cannot simply do something like
```ruby
write_text('hello', row: 1, col: 15)
write_text('ruby', row: 2, col: 5)
write_text('world', row: 3, col: 10)
```
But move cursor and write
```ruby
move_cursor(row: 1, col: 15) # cursor is at [1, 15]
write_to_stdout('hello') # cursor is at [1, 15 + 6]
move_cursor(row: 2, col: 5) # cursor is at [2, 5]
write_to_stdout('ruby') # cursor is at [2, 5 + 4]
move_cursor(row: 3, col: 10) # cursor is at [3, 10]
write_to_stdout('world') # cursor is at [3, 10 + 5]
move_cursor(row: initial_cursor_row, col: initial_cursor_col)
```
This is why I need to move cursor many times.

Reply:
Do you mean we need to move the cursor to the place that we want to start writing next? Like the cursor is our interface to write things to the terminal?

Reply:
Yes, that is what reline, emacs and vim are doing.

Reply:
I had no idea it works like that! Thank you for explaining it to me patiently. The slow-mo videos are also helpful 🙏